cql3/expr/expression.cc

```diff
@@ -2417,5 +2417,24 @@ bool is_empty_restriction(const expression& e) {
     return find_in_expression<column_value>(e, [](const auto&) {return true;}) == nullptr;
 }
 
+sstring get_columns_in_commons(const expression& a, const expression& b) {
+    std::vector<const column_definition*> ours = get_sorted_column_defs(a);
+    std::vector<const column_definition*> theirs = get_sorted_column_defs(b);
+
+    std::sort(ours.begin(), ours.end());
+    std::sort(theirs.begin(), theirs.end());
+    std::vector<const column_definition*> common;
+    std::set_intersection(ours.begin(), ours.end(), theirs.begin(), theirs.end(), std::back_inserter(common));
+
+    sstring str;
+    for (auto&& c : common) {
+        if (!str.empty()) {
+            str += " ,";
+        }
+        str += c->name_as_text();
+    }
+    return str;
```

Comment:
nitpick: it looks like you reinvented fmt::join (https://fmt.dev/latest/api.html) from scratch

Reply:
I just copied the original `columns_in_commons` without thinking much.

I tried using `fmt::join`, but it looks like `sstring` is not formattable using `fmt::format` 0_o.

Something like this fails with an error message:

```c++
return fmt::format("{}",
    fmt::join(
        common | boost::adaptors::transformed([](const column_definition* cdef) { return cdef->name_as_text();}),
        " ,"));
```

But casting lambda result to `std::string` makes it work:
```c++
return fmt::format("{}",
        fmt::join(
            common | boost::adaptors::transformed([](const column_definition* cdef) -> const std::string { return cdef->name_as_text();}),
            " ,"));
```

I don't want to implement `formattable` for `sstring` in this PR. There is enough going on already.
Let's leave it for a separate PR.

Here's the error message:

```

/usr/include/fmt/core.h:1422:3: error: static_assert failed due to requirement 'fmt::formattable<fmt::arg_join<boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *, std::allocator<const column_definition *>>>, boost::use_default, boost::use_default>, boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *, std::allocator<const column_definition *>>>, boost::use_default, boost::use_default>, char>>()' "Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt"
  static_assert(
  ^
/usr/include/fmt/core.h:1438:10: note: in instantiation of function template specialization 'fmt::detail::check<fmt::arg_join<boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, char>>' requested here
  return check<T>(arg_mapper<Context>().map(val));
         ^
/usr/include/fmt/core.h:1587:23: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::basic_format_context<fmt::detail::buffer_appender<char>, char>, fmt::detail::type::custom_type, fmt::arg_join<boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, char>, 0>' requested here
        data_{detail::make_arg<
                      ^
/usr/include/fmt/core.h:1626:10: note: in instantiation of member function 'fmt::format_arg_store<fmt::basic_format_context<fmt::detail::buffer_appender<char>, char>, fmt::arg_join<boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, char>>::format_arg_store' requested here
  return {args...};
         ^
/usr/include/fmt/core.h:2075:28: note: in instantiation of function template specialization 'fmt::make_args_checked<fmt::arg_join<boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, char>, char [3], char>' requested here
  const auto& vargs = fmt::make_args_checked<Args...>(format_str, args...);
                           ^
cql3/expr/expression.cc:2425:17: note: in instantiation of function template specialization 'fmt::format<char [3], fmt::arg_join<boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, boost::iterators::transform_iterator<(lambda at cql3/expr/expression.cc:2427:51), __gnu_cxx::__normal_iterator<const column_definition **, std::vector<const column_definition *>>>, char>, char>' requested here
    return fmt::format("{}",
                ^
1 error generated.
ninja: build stopped: subcommand failed.

```